File: 9M2W/minjeeki/boj_1987.py
import sys
from collections import deque


data = sys.stdin.read().strip().split("\n")
R, C = map(int, data[0].split())
table = [i for i in data[1:]]
visited = [[False]*C for _ in range(R)]
q = deque([[(0,0),table[0][0]]])
visited[0][0] = True
dx = [1, -1, 0, 0]
dy = [0, 0, 1, -1]
answer = 0


def dfs(x, y, history):
    global answer
    if len(history) > answer:
        answer = len(history)

    visited[x][y] = True

    for i in range(4):
        nx = x + dx[i]
        ny = y + dy[i]
        if 0 <= nx < R and 0 <= ny < C and table[nx][ny] not in history:
            if not visited[nx][ny]:
                dfs(nx, ny, history + table[nx][ny])
    
    visited[x][y] = False

dfs(0, 0, table[0][0])
print(answer)

File: 9M2W/minjeeki/boj_1987.js
const fs = require('fs');

const START = [0, 0];
const dx = [0, 0, 1, -1];
const dy = [1, -1, 0, 0];

function backTracking(curX, curY, depth) {
    const remainingAlphabets = 26 - visited.reduce((sum, v) => sum + (v ? 1 : 0), 0);
    if (depth + remainingAlphabets <= answer) {
        return;
    }
    
    if (depth > answer) {
        answer = depth;
    }
    if (answer === 26) {
        return;
    }

    for (let k = 0; k < 4; k++) {
        const nx = curX + dx[k];
        const ny = curY + dy[k];
        if (0 <= nx && nx < R && 0 <= ny && ny < C) {
            const alphaIdx = board[nx][ny].charCodeAt(0) - 'A'.charCodeAt(0);
            if (!visited[alphaIdx]) {
                visited[alphaIdx] = true;
                backTracking(nx, ny, depth + 1);
                visited[alphaIdx] = false;
            }
        }
    }
}

const data = fs.readFileSync('/dev/stdin', 'utf8').trim().split('\n');
const [R, C] = data[0].split(' ').map(Number);
const board = data.slice(1).map(line => line.trim());

let answer = 0;
let visited = new Array(26).fill(false);
visited[board[0][0].charCodeAt(0) - 'A'.charCodeAt(0)] = true;
backTracking(START[0], START[1], 1);

console.log(answer);
